### DISCUSSION TITLE
Problem with update & fast consecutive requests

### DISCUSSION THREAD

**Discussion by felixkpunkt** [prisma/prisma#13422] [Q&A]
Hi!

I'm currently using Prisma with a PostgreSQL database & NEXT.js.
Unfortunetely I'm stuck with a weird problem I can't figure out. In short: I have the following code which is executed on a /refresh endpoint. It takes care of reading, validating and creating a new refresh token for a JWT based login.

The problem is a difference between ```newRefreshToken``` in the database and the cookie in the browser if I hit F5 a few times in a row.

```typescript
await prisma.user.update({
      where: { id: user.id },
      data: { refreshToken: newRefreshToken },
    });
```
doest not seem to update the refresh token in the user record if there are multiple requests incoming and therefore it fails because the code is looking for the new refresh token:
```typescript
const user = await prisma.user.findUnique({
    where: {
      refreshToken,
    },
  });
```

```typescript
export default async function refreshRoute(
  req: NextApiRequest,
  res: NextApiResponse
) {
  if (req.method !== 'GET') {
    return res.status(405).json({ message: 'Method not allowed!' });
  }

  const { refresh_token: refreshToken } = cookie.parse(
    req.headers.cookie ?? ''
  );

  if (undefined === refreshToken) {
    return res.status(401).json({ message: 'refresh_token_is_missing' });
  }

  let decoded;
  try {
    decoded = jwt.verify(
      refreshToken,
      process.env.REFRESH_TOKEN_SECRET ?? ''
    ) as TokenPayload;
  } catch (error) {
    return res
      .setHeader('Set-Cookie', createExpiredCookie())
      .status(401)
      .json({ message: 'invalid_refresh_token' });
  }

  const user = await prisma.user.findUnique({
    where: {
      refreshToken,
    },
  });

  if (null === user) {
    return res
      .setHeader('Set-Cookie', createExpiredCookie())
      .status(401)
      .json({ message: 'invalid_session' });
  }

  const payload: TokenPayload = {
    id: user.id,
  };
  const accessToken = generateAccessToken(payload);
  const newRefreshToken = generateRefreshToken(payload);

  try {
    await prisma.user.update({
      where: { id: user.id },
      data: { refreshToken: newRefreshToken },
    });
  } catch (error) {
    return res
      .setHeader('Set-Cookie', createExpiredCookie())
      .status(500)
      .json({ message: 'error_creating_refresh_token' });
  }

  return res
    .setHeader(
      'Set-Cookie',
      cookie.serialize('refresh_token', newRefreshToken, {
        httpOnly: true,
        maxAge: 60 * 60 * 24 * 7,
        sameSite: true,
        path: '/',
      })
    )
    .status(200)
    .json({ message: 'refresh_successful', accessToken });
}
```

Best regards!

**Comment by nurul3101** [comment]
Hey @knojector 👋 
Are you on the latest `@prisma/client` version i.e. `3.14.0`?
Can you try enabling [logging](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/logging) to validate if the database query is being sent to the database?

**Comment by felixkpunkt** [comment]
Hi @nurul3101 
thank you for your response!

Yes, I'm using the latest version. I enabled logging and saw in the logged transactions that the SELECT query for the user by token was executed before the previous UPDATE query (from the previous request) had finished updating the token. So this is how the difference happened - at least I think. I solved the problem by issuing a new token only once per minute.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
